h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java

```diff
@@ -0,0 +1,343 @@
+      final double scale = parms._scale == null ? 1.0 : parms._scale[index];
+      final GAMParameters.BSType splineType = parms._bs[index];
+      final int frameIndex = index;
+      final String[] newColNames = new String[numKnots];
+      for (int colIndex = 0; colIndex < numKnots; colIndex++) {
+        newColNames[colIndex] = parms._gam_X[index] + "_" + splineType.toString() + "_" + colIndex;
+      }
+      gamColnames[frameIndex] = new String[numKnots];
+      gamColnamesCenter[frameIndex] = new String[numKnotsM1];
+      knotsMat[frameIndex] = new double[numKnots];
+      System.arraycopy(newColNames, 0, gamColnames[frameIndex], 0, numKnots);
+      generateGamColumn[frameIndex] = new RecursiveAction() {
+        @Override
+        protected void compute() {
+          GenerateGamMatrixOneColumn genOneGamCol = new GenerateGamMatrixOneColumn(splineType, numKnots, null, predictVec,
+                  parms._standardize, centerGAM, scale).doAll(numKnots, Vec.T_NUM, predictVec);
+          Frame oneAugmentedColumn = genOneGamCol.outputFrame(Key.make(), newColNames,
+                  null);
+          gamFrameKeys[frameIndex] = oneAugmentedColumn._key;
+          DKV.put(oneAugmentedColumn);
+          if (modelBuilding) {  // z and penalty matrices are only needed during model building
+            if (centerGAM) {  // calculate z transpose
+              Frame oneAugmentedColumnCenter = genOneGamCol.outputFrame(Key.make(), newColNames,
+                      null);
+              oneAugmentedColumnCenter = genOneGamCol.de_centralize_frame(oneAugmentedColumnCenter,
+                      predictVec.name(0) + "_" + splineType.toString() + "_decenter_", parms);
+              GamUtils.copy2DArray(genOneGamCol._ZTransp, zTranspose[frameIndex]); // copy transpose(Z)
+              double[][] transformedPenalty = ArrayUtils.multArrArr(ArrayUtils.multArrArr(genOneGamCol._ZTransp,
+                      genOneGamCol._penaltyMat), ArrayUtils.transpose(genOneGamCol._ZTransp));  // transform penalty as zt*S*z
+              GamUtils.copy2DArray(transformedPenalty, penalty_mat[frameIndex]);
+
+              gamFrameKeysCenter[frameIndex] = oneAugmentedColumnCenter._key;
+              DKV.put(oneAugmentedColumnCenter);
+              System.arraycopy(oneAugmentedColumnCenter.names(), 0, gamColnamesCenter[frameIndex], 0,
+                      numKnotsM1);
+            } else
+              GamUtils.copy2DArray(genOneGamCol._penaltyMat, penalty_mat[frameIndex]); // copy penalty matrix
+            GamUtils.copy2DArray(genOneGamCol._bInvD, binvD[frameIndex]);
+            numKnotsMat[frameIndex] = genOneGamCol._numKnots;
+            System.arraycopy(genOneGamCol._knots, 0, knotsMat[frameIndex], 0, numKnots);
+          }
+        }
+      };
+    }
+    ForkJoinTask.invokeAll(generateGamColumn);
+  }
+  
+  public static Frame buildGamFrame(int numGamFrame, Key<Frame>[] gamFramesKey, Frame _train, String response_column) {
```

Comment:
This will add to the _train frame basis function expansions for each column specified in _parms._gam_X